File: sdks/go/coupons.go
package settlekit

import (
	"context"
	"net/http"
)

// CreateCouponInput is the body for CreateCoupon (POST /v1/coupons). Discount
// is required; populate the field matching Discount.Type.
type CreateCouponInput struct {
	Code                string         `json:"code"`
	Name                string         `json:"name,omitempty"`
	Discount            CouponDiscount `json:"discount"`
	Status              string         `json:"status,omitempty"`
	StartsAt            string         `json:"startsAt,omitempty"`
	ExpiresAt           string         `json:"expiresAt,omitempty"`
	MaxRedemptions      int            `json:"maxRedemptions,omitempty"`
	PerCustomerLimit    int            `json:"perCustomerLimit,omitempty"`
	MinSubtotal         string         `json:"minSubtotal,omitempty"`
	AppliesToProductIDs []string       `json:"appliesToProductIds,omitempty"`
}

// CreateCoupon creates a discount coupon. POST /v1/coupons.
func (c *Client) CreateCoupon(ctx context.Context, in CreateCouponInput) (*Coupon, error) {
	var out Coupon
	if err := c.do(ctx, http.MethodPost, "/v1/coupons", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListCoupons returns all coupons. GET /v1/coupons.
func (c *Client) ListCoupons(ctx context.Context) ([]Coupon, error) {
	var out []Coupon
	if err := c.do(ctx, http.MethodGet, "/v1/coupons", nil, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// GetCoupon fetches a coupon by code. GET /v1/coupons/:code.
func (c *Client) GetCoupon(ctx context.Context, code string) (*Coupon, error) {
	var out Coupon
	if err := c.do(ctx, http.MethodGet, "/v1/coupons/"+code, nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ApplyCouponInput is the body for ValidateCoupon / RedeemCoupon.
type ApplyCouponInput struct {
	Subtotal   string `json:"subtotal"`
	CustomerID string `json:"customerId,omitempty"`
}

// ApplyCouponResult is the outcome of applying a coupon to a subtotal.
type ApplyCouponResult struct {
	OK            bool   `json:"ok"`
	Discount      Money  `json:"discount"`
	Total         Money  `json:"total"`
	FreeTrialDays int    `json:"freeTrialDays,omitempty"`
	Reason        string `json:"reason,omitempty"`
}

// ValidateCoupon dry-runs a coupon against a subtotal without mutating its
// redemption count. POST /v1/coupons/:code/validate.
func (c *Client) ValidateCoupon(ctx context.Context, code string, in ApplyCouponInput) (*ApplyCouponResult, error) {
	var out ApplyCouponResult
	if err := c.do(ctx, http.MethodPost, "/v1/coupons/"+code+"/validate", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// RedeemCoupon redeems a coupon against a subtotal, mutating its redemption
// count. POST /v1/coupons/:code/redeem.
func (c *Client) RedeemCoupon(ctx context.Context, code string, in ApplyCouponInput) (*ApplyCouponResult, error) {
	var out ApplyCouponResult
	if err := c.do(ctx, http.MethodPost, "/v1/coupons/"+code+"/redeem", in, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: sdks/rust/src/resources/coupons.rs
//! Coupon (discount) resource (`/v1/coupons`).

use reqwest::Method;
use serde::{Deserialize, Serialize};

use crate::{Client, Coupon, CouponDiscount, Result};

/// Accessor for coupon operations.
#[derive(Debug)]
pub struct Coupons<'a> {
    client: &'a Client,
}

/// Body for creating a coupon (`POST /v1/coupons`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateCoupon {
    pub code: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    pub discount: CouponDiscount,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub starts_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_redemptions: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub per_customer_limit: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub min_subtotal: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub applies_to_product_ids: Option<Vec<String>>,
}

/// Body for validating/redeeming a coupon against a subtotal.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ApplyCoupon {
    pub subtotal: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub customer_id: Option<String>,
}

/// Outcome of validating or redeeming a coupon. Fields are captured loosely
/// since the engine returns a result/outcome shape.
#[derive(Debug, Clone, Deserialize)]
#[serde(transparent)]
pub struct CouponApplyResult(pub serde_json::Value);

impl<'a> Coupons<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Coupons { client }
    }

    /// Create a coupon.
    pub async fn create(&self, body: &CreateCoupon) -> Result<Coupon> {
        self.client
            .request(Method::POST, "/v1/coupons", Some(body))
            .await
    }

    /// List coupons.
    pub async fn list(&self) -> Result<Vec<Coupon>> {
        self.client
            .request_no_body(Method::GET, "/v1/coupons")
            .await
    }

    /// Fetch a coupon by code.
    pub async fn get(&self, code: &str) -> Result<Coupon> {
        self.client
            .request_no_body(Method::GET, &format!("/v1/coupons/{code}"))
            .await
    }

    /// Dry-run apply a coupon against a subtotal (no mutation).
    pub async fn validate(&self, code: &str, body: &ApplyCoupon) -> Result<CouponApplyResult> {
        self.client
            .request(
                Method::POST,
                &format!("/v1/coupons/{code}/validate"),
                Some(body),
            )
            .await
    }

    /// Redeem a coupon against a subtotal (increments its redemption count).
    pub async fn redeem(&self, code: &str, body: &ApplyCoupon) -> Result<CouponApplyResult> {
        self.client
            .request(
                Method::POST,
                &format!("/v1/coupons/{code}/redeem"),
                Some(body),
            )
            .await
    }
}
